### PR TITLE
`Exam mode`: Improve usability and readability of results summary

### PR DESCRIPTION
**Pull request by florian-glombik**
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
       Did not find a better alternative to "Prüfungsteilnehmer:in"
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style, wherever the method name is not self-explanatory.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Resolves https://github.com/ls1intum/Artemis/issues/7203

### Description
<!-- Describe your changes in detail -->

#### Exam Mode Testing
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

Prerequisites:
- 1 Instructor
- 1 Students _(You can also view the Student Submission as instructor, then you do not need to use a student)_
- 1 Exam with all exercise types _(Programming, Quiz, Modelling, FileUpload, Text)_

1. student: participate in the exam, see that the confirmation screen appearance has changed
2. student: view the ungraded summary directly after participating in an exam
3. instructor: start and participate in a test run exam -> the start page and summary are relevant
4. student: view the summary with a graded participation
    a) the general information should be easier to understand
    b) the result overview should link to the exercises (onClick -> scrolling to exercise)
    c) the grading keys should be collapsable (and no links to other pages)
    d) readable in darkmode
    e) the grading scale button is not visible if no grading scale exists
    d) the bonus grading scale button is not visible of no bonus grading scale exists
5. instructor: the grading key table is still displayed properly when accessed as page


### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |


just refactored the files with 0% coverage, added tests for exam-result-overview where more functionality was added
-->

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam-result-overview.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/result-overview/exam-result-overview.component.ts.html) | 94.23% | ✅ |
| [participation.utils.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/participation/participation.utils.ts.html) | 71.11% |  |
| [updating-result.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/result/updating-result.component.ts.html) | 98.24% |  |
| [grading-key-helper.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/grading-system/grading-key-overview/grading-key-helper.ts.html) | 100% | ✅ |
| [grading-key-overview.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/grading-system/grading-key-overview/grading-key-overview.component.ts.html) | 100% | ✅ |
| [grading-key-table.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5435/latest/artifact/shared/Coverage-Report-Client-Tests/app/grading-system/grading-key-overview/grading-key/grading-key-table.component.ts.html) | 98.03% |  |






### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

| Before                                                                                                                                        | After                                                                                                                                                                                                                                                                                         |
|-----------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <img width="1248" alt="exam cover page after" src="https://github.com/ls1intum/Artemis/assets/63976129/aea7e3d7-1b1c-4705-a528-45e962a4f54e"> | <img width="1248" alt="exam cover page before" src="https://github.com/ls1intum/Artemis/assets/63976129/f583aff8-a4ab-4d6c-a19b-41eed58142fd">                                                                                                                                                |
| <img width="1248" alt="exam summary before" src="https://github.com/ls1intum/Artemis/assets/63976129/a4f0a509-ad54-462a-9558-c5d5ce29cc3d">   | <img width="1248" alt="exam cover page before" src="https://github.com/ls1intum/Artemis/assets/63976129/3ecadcf7-122b-45d3-9623-56cfc4c44fac"> <img width="1248" alt="exam cover page before" src="https://github.com/ls1intum/Artemis/assets/63976129/13fe9b89-b1aa-413c-b62e-be83df402c13"> |

### CODE REVIEW

**Review by julian-christl — DISMISSED** [review]
Tested on TS1 looks good. Issues I found:
- Bonus Grading Key Button should not be visible if there are no keys
- Bonuus Grading Keys are right aligned instead of left
- For the preview page: I recommend reading duration, points and exercises for the mentioned reasons

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS2. We discussed some optional improvements that can be part of a followup.

File: src/main/webapp/app/exercises/shared/result/updating-result.component.ts [hunk 1]
```diff
@@ -172,3 +165,14 @@ export class UpdatingResultComponent implements OnChanges, OnDestroy {
...
+export function getLatestResultOfStudentParticipation(participation: StudentParticipation, showUngradedResults: boolean): Result | undefined {
+    // Sort participation results by completionDate desc.
+    if (participation.results) {
+        participation.results = _orderBy(participation.results, 'completionDate', 'desc');
+    }
+    // The latest result is the first rated result in the sorted array (=newest) or any result if the option is active to show ungraded results.
+    const latestResult = participation.results?.find(({ rated }) => showUngradedResults || rated === true);
+    // Make sure that the participation result is connected to the newest result.
+    return latestResult ? { ...latestResult, participation: participation } : undefined;
+}
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/result/updating-result.component.ts:178]: Can you move this into a shard place? This method should be useful for many other places as well.

Comment (florian-glombik) [reply src/main/webapp/app/exercises/shared/result/updating-result.component.ts]: I moved it to the `participation.utils.ts` 👍

**Review by julian-christl — DISMISSED** [review]
Tested on TS2, all looks good now. Thanks for implementing the changes!

**Comment by florian-glombik** [comment]
The misleading description within the bonus grading key table is known and will be addressed in a follow-up PR - this issue does already exist on develop.

image

**Review by julian-christl — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by Strohgelaender — APPROVED** [review]
re-tested on TS2, I did not find any issues and the new look is quite improved.

In a followup we might improve this tooltip, I think it should be much simpler for students (e.g. just ' You did not participate in this course').
grafik

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.scss [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+.rotate-icon.rotated {
+    transform: rotate(90deg);
+}
+
+.icon-container {
+    display: inline-block;
+    vertical-align: middle;
+    horiz-align: center;
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.scss:49]: I don't think that exists? Maybe look it up?

File: src/main/webapp/app/exam/participate/general-information/exam-general-information.component.ts [hunk 3]
```diff
@@ -5,12 +5,18 @@ import { endTime, getAdditionalWorkingTime, isExamOverMultipleDays, normalWorkin
...
+    styleUrls: ['./exam-general-information.component.scss'],
+    templateUrl: './exam-general-information.component.html',
 })
-export class ExamInformationComponent implements OnInit {
+export class ExamGeneralInformationComponent implements OnInit {
     @Input() exam: Exam;
     @Input() studentExam: StudentExam;
+    @Input() reviewIsOpen?: boolean = false;
+
+    /** The exam cover will contain e.g. the number of exercises which is hidden in the exam summary as
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exam/participate/general-information/exam-general-information.component.ts:17]: Doc Style

File: src/test/javascript/spec/component/grading-system/grading-key-table.component.spec.ts
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/grading-system/grading-key-table.component.spec.ts:1]: We usually don't do that: `// eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain`

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
I've some suggestions.

**Review by YeetTheFirst21 — APPROVED** [review]
Retested on TS2,

looks really nice as long as the student name or course name is not very long

**Review by rriyaldhi — APPROVED** [review]

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.scss (hunk 2 above)
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.scss:49]: Addressed in https://github.com/ls1intum/Artemis/pull/7281/commits/24bd8bf73203f2a557c885dc7956ee78a5e8f7f1

File: src/main/webapp/app/exam/participate/general-information/exam-general-information.component.ts (hunk 3 above)
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/general-information/exam-general-information.component.ts:17]: Addressed in https://github.com/ls1intum/Artemis/pull/7281/commits/24bd8bf73203f2a557c885dc7956ee78a5e8f7f1

**Pull request merged**